pkg/sfu/videolayerselector/dependencydescriptor.go

```diff
@@ -332,6 +332,17 @@ func (d *DependencyDescriptor) CheckSync() (locked bool, layer int32) {
 		return false, layer
 	}
 
+	allBroken := true
```

Comment:
There is a potential race condition here as d.chains is not protected by lock. I think the old logic may have a problem to only check the request spatial layer, we can check if there is a decodable target that is lesser or equal to the request one like this:
```
for _, dt := range d.decodeTargets {
    if dt.Active && dt.Layer.Spatial < requestLayer && dt.Valid() {
        return true,layer
    }
}
```

Reply:
It is actually protected by `forwarder` lock, but let us not rely on that. Is it okay to protect with the `decodeTargets` lock?

Will make the change to check for <= (should be <= and not just < requestLayer, right?)

Reply:
my bad, it is protected by forwarder lock so no race condition here. I would prefer to check decode target instead of decode chain as the target represents a special layer. If the `CheckSync` means there is a decodeable layer that is lesser or equal than the `RequestSpatial`, I think we don't need add extra check but just change the == to <= .

Reply:
oh, you mean to say that all chains broken is  not needed? And just check for decodable? 

I think we can then change to `targetLayer`. The `requestSpatial` is an odd one. As `forwarder` allows overshooting (to deal with some clients that do not do layers properly), the target can be high, but request layer spatial refers to which layer the PLI request should be made on as all layers up to the target layer may not be available.

For SVC + DD, I think that is not important as it is single layer. So, we can change it to if decode target <= targetLayer is available, it should be good.

Let me make the change and attach diff here for your review.

Reply:
@cnderrauber The diff
```
diff --git a/pkg/sfu/videolayerselector/dependencydescriptor.go b/pkg/sfu/videolayerselector/dependencydescriptor.go
index 1fc8e5f3..63ba36f7 100644
--- a/pkg/sfu/videolayerselector/dependencydescriptor.go
+++ b/pkg/sfu/videolayerselector/dependencydescriptor.go
@@ -332,21 +332,10 @@ func (d *DependencyDescriptor) CheckSync() (locked bool, layer int32) {
                return false, layer
        }
 
-       allBroken := true
-       for _, chain := range d.chains {
-               if !chain.Broken() {
-                       allBroken = false
-                       break
-               }
-       }
-       if allBroken {
-               return false, layer
-       }
-
        d.decodeTargetsLock.RLock()
        defer d.decodeTargetsLock.RUnlock()
        for _, dt := range d.decodeTargets {
-               if dt.Active() && dt.Layer.Spatial == layer && dt.Valid() {
+               if dt.Active() && dt.Layer.Spatial <= d.GetTarget().Spatial && dt.Valid() {
                        d.logger.Debugw(fmt.Sprintf("checking sync, matching decode target, layer: %d, dt: %s, dts: %+v", layer, dt, d.decodeTargets))
                        return true, layer
                }
```

And the full function
```
func (d *DependencyDescriptor) CheckSync() (locked bool, layer int32) {
	layer = d.GetRequestSpatial()
	if !d.currentLayer.IsValid() {
		// always declare not locked when trying to resume from nothing
		return false, layer
	}

	d.decodeTargetsLock.RLock()
	defer d.decodeTargetsLock.RUnlock()
	for _, dt := range d.decodeTargets {
		if dt.Active() && dt.Layer.Spatial <= d.GetTarget().Spatial && dt.Valid() {
			d.logger.Debugw(fmt.Sprintf("checking sync, matching decode target, layer: %d, dt: %s, dts: %+v", layer, dt, d.decodeTargets))
			return true, layer
		}
	}

	return false, layer
}
```

How does that look?

Reply:
Thanks! It looks good to me and make the request layer more clearly to me.

Reply:
Oops, I merged before coming that change. I will make another PR. Sorry.